src/passes/OptimizeCasts.cpp

```diff
@@ -102,6 +120,256 @@ namespace wasm {
+    auto iter = refAsReplacement.find(moveableExpressions.front());
+    if (iter == refAsReplacement.end()) {
+      // Since RefAs ops are orthogonal to each other, we currently just
+      // select the first RefAs we come accross which is moveable.
+      refAsReplacement.emplace(moveableExpressions.front(), curr);
+      if (fallthrough == moveableExpressions.front()) {
+        redundantRefAsReplacement.insert(fallthrough);
+      }
+    }
+  }
+
+  void visitRefCast(RefCast* curr, LocalGet* fallthrough) {
+    auto iter = refCastReplacement.find(moveableExpressions.front());
+    if (iter == refCastReplacement.end()) {
+      refCastReplacement.emplace(moveableExpressions.front(), curr);
+      if (fallthrough == moveableExpressions.front()) {
+        redundantRefCastReplacement.insert(fallthrough);
+      }
+    } else if (curr->type != iter->second->type &&
+               Type::isSubType(curr->type, iter->second->type)) {
+      iter->second = curr;
+      if (fallthrough != iter->first) {
+        redundantRefCastReplacement.erase(iter->first);
+      }
+    }
+  }
+
+  void removeRedundantReplacements() {
+    for (auto& redundantRefAs : redundantRefAsReplacement) {
+      refAsReplacement.erase(redundantRefAs);
+    }
+    for (auto& redundantRefCast : redundantRefCastReplacement) {
+      refCastReplacement.erase(redundantRefCast);
+    }
+  }
+};
+
+// Find the "best" cast to move earlier to another local.gets
+struct EarlyCastFinder
+  : public LinearExecutionWalker<EarlyCastFinder,
+                                 UnifiedExpressionVisitor<EarlyCastFinder>> {
+  PassOptions options;
+
+  // For each local index, we have an searcher to move casts for it.
+  std::unordered_map<Index, IndexEarlyCastSearcher> castMoveSearchers;
+
+  // Maps local.gets to the "best" RefAs to move to it.
+  std::unordered_map<Expression*, RefAs*> refAsToApply;
+
+  // Maps local.gets to the "best" RefCast to move to it.
+  std::unordered_map<Expression*, RefCast*> refCastToApply;
+
+  static void doNoteNonLinear(EarlyCastFinder* self, Expression**) {
+    self->castMoveSearchers.clear();
+  }
+
+  void visitExpression(Expression* curr) {
+    for (auto& indexSearcher : castMoveSearchers) {
+      indexSearcher.second.visitExpression(curr);
+    }
+  }
+
+  void visitLocalSet(LocalSet* curr) {
+    for (auto& indexSearcher : castMoveSearchers) {
+      indexSearcher.second.visitExpression(curr);
+    }
+    auto iter = castMoveSearchers.find(curr->index);
+    if (iter != castMoveSearchers.end()) {
+      iter->second.visitLocalSet(curr);
+    }
+  }
+
+  void visitLocalGet(LocalGet* curr) {
+    for (auto& indexSearcher : castMoveSearchers) {
+      indexSearcher.second.visitExpression(curr);
+    }
+    auto iter = castMoveSearchers.find(curr->index);
+    if (iter == castMoveSearchers.end()) {
+      RefCast testRefCast(getModule()->allocator);
+      EffectAnalyzer efa(options, *getModule());
+      efa.visit(&testRefCast);
```

Comment:
Reading this code, it's not obvious to me why we compute the effects of a RefCast when we are processing a LocalGet, so I think that is worth a comment at least.

Reply:
It seems like this was something from an earlier iteration that I forgot to delete. We don't need to compute the effects of a RefCast here. I will remove it in the next commit.